Задание №3: веб-приложение

RealMadSquirrel opened the pull request:
Реализовано:
### Восстановление пароля
- переход на страницу восстановления пароля по кнопке «Восстановить пароль»,
- ввод почты и клик по кнопке «Восстановить»,
- клик по кнопке показать/скрыть пароль делает поле активным — подсвечивает его.
### Личный кабинет
- переход по клику на «Личный кабинет»,
- переход в раздел «История заказов»,
- выход из аккаунта.

### Проверка основного функционала
- переход по клику на «Конструктор»,
- переход по клику на «Лента заказов»,
- если кликнуть на ингредиент, появится всплывающее окно с деталями,
- всплывающее окно закрывается кликом по крестику,
- при добавлении ингредиента в заказ счётчик этого ингридиента увеличивается,
- залогиненный пользователь может оформить заказ.

### Раздел «Лента заказов»
- если кликнуть на заказ, откроется всплывающее окно с деталями,
- заказы пользователя из раздела «История заказов» отображаются на странице «Лента заказов»,
- при создании нового заказа счётчик Выполнено за всё время увеличивается,
- при создании нового заказа счётчик Выполнено за сегодня увеличивается,
- после оформления заказа его номер появляется в разделе В работе.

Commit:
Реализованы проверки на восстановление пароля:
 - переход на страницу восстановления пароля по кнопке «Восстановить пароль»,
 - ввод почты и клик по кнопке «Восстановить»,
 - клик по кнопке показать/скрыть пароль делает поле активным — подсвечивает его.

Commit:
Реализованы проверки на Личный кабинет и часть проверок основного функционала:
 - переход по клику на «Личный кабинет»,
 - переход в раздел «История заказов»,
 - выход из аккаунта.

 - переход по клику на «Конструктор»,
 - переход по клику на «Лента заказов»,
 - если кликнуть на ингредиент, появится всплывающее окно с деталями,
всплывающее окно закрывается кликом по крестику,

Commit:
Добавлены проверки:

Проверка основного функционала

-переход по клику на «Конструктор»,
-переход по клику на «Лента заказов»,
-если кликнуть на ингредиент, появится всплывающее окно с деталями,
всплывающее окно закрывается кликом по крестику,
-при добавлении ингредиента в заказ счётчик этого ингридиента увеличивается,
-залогиненный пользователь может оформить заказ.

Раздел «Лента заказов»

-если кликнуть на заказ, откроется всплывающее окно с деталями,
-заказы пользователя из раздела «История заказов» отображаются на странице «Лента заказов»,
-при создании нового заказа счётчик Выполнено за всё время увеличивается,
-при создании нового заказа счётчик Выполнено за сегодня увеличивается,
после оформления заказа его номер появляется в разделе В работе.

AlexKitIt commented on conftest.py:
```diff
@@ -0,0 +1,31 @@
+    elif request.param == 'firefox':
+        browser = webdriver.Firefox()
+        browser.maximize_window()
+    #browser.get(Urls.BASE_URL)
```
Нужно исправить: закомменченного кода быть не должно

RealMadSquirrel replied:
Убрала

AlexKitIt commented on conftest.py:
```diff
@@ -0,0 +1,31 @@
+@pytest.fixture(scope='function')
+def create_creds_user():
+    creds = helper.UserFactory.register_new_user_and_return_load()
+    return creds
```
коммент из 2 проекта

RealMadSquirrel replied:
Поправила

AlexKitIt commented on locators/main_page_locators.py:
```diff
@@ -0,0 +1,18 @@
+class MainPageLocators:
+    PERSONAL_BUTTON = (By.XPATH, "//a[@href = '/account']") # ссылка на Личный Кабинет
+    PROFILE_TITLE = (By.XPATH, "//a[@href = '/account/profile']")
+    LENTA_ORDERS = (By.XPATH, "//h1[text() = 'Лента заказов']")
+    LENTA_BUTTON = (By.XPATH, "//p[text() = 'Лента Заказов']/parent::a")
```
Нужно исправить: здесь и далее: элемент всегда располагается на одной странице, он не копипастится из пейджи в пейджу, а для работы с ним в тесте создается объект конкретной страницы, происходит взаимодействие, потом создается другой объект страницы и взаимодействие с другими элементами

RealMadSquirrel replied:
Поправила https://github.com/RealMadSquirrel/Diplom_3/pull/1/commits/dc57ead3a74724e066dbb53073f3854e1b1022b1

AlexKitIt commented on pages/base_page.py:
```diff
@@ -0,0 +1,98 @@
+
+    @allure.step('Ожидаем и находим элемент')
+    def wait_and_find_element(self, locator):
+        WebDriverWait(self.driver, 20).until(expected_conditions.visibility_of_element_located(locator))
```
Нужно исправить: здесь и далее: метод описан в check_exists_by_xpath необходимо переиспользовать а не копипастить

RealMadSquirrel replied:
Поправила, заменила check_exists_by_xpath на wait_and_find_element.
Ссылка на комит - https://github.com/RealMadSquirrel/Diplom_3/pull/1/commits/5480db73482872a6b0e30179cd74f1f0d362e1b0

AlexKitIt commented on pages/login_page.py:
```diff
@@ -0,0 +1,32 @@
+    def click_button_go(self):
+        self.wait_and_find_element(LoginPageLocators.BUTTON_GO)
+        self.click_to_element_with_js(LoginPageLocators.BUTTON_GO)
+        return MainPage(self.driver)
+
+
+    @allure.step('Кликаем по кнопке «Восстановить пароль»')
+    def click_restore_password(self):
+        self.wait_and_find_element(LoginPageLocators.FORGOT_PASSWORD)
+        self.click_to_element_with_js(LoginPageLocators.FORGOT_PASSWORD)
+        return ForgotPasswordPage(self.driver)
```
Нужно исправить: здесь и далее: не стоит создавать объект одной страницы в методах другой, это нарушает принцип инкапсуляции логики пейджей

RealMadSquirrel replied:
Убрала

AlexKitIt commented on pages/main_page.py:
```diff
@@ -0,0 +1,25 @@
+
+
+class MainPage(BasePage):
+    CURRENT_ACCOUNT_LOGIN = (By.XPATH, "//p[@class = 'header__user']")
```
Нужно исправить: здесь и далее: все локаторы хранятся отдельно от тестов в пакете locators

RealMadSquirrel replied:
Убрала этот локатор

AlexKitIt commented on pages/orders_lenta_page.py:
```diff
@@ -0,0 +1,12 @@
+
+    @allure.step('Кликаем на кнопку Личный кабинет')
+    def click_personal_account(self):
+        self.wait_and_click_to_element(MainPageLocators.PERSONAL_BUTTON)
```
Нужно исправить: здесь и далее: в методах пейджи описываются только методы работы с элементами на конкретной пейдже, работы с элементами с других страниц быть не должно Это OrderLentaPage а локатор с MainPage

RealMadSquirrel replied:
Поправила.

AlexKitIt commented on tests/test_main_functionality.py:
```diff
@@ -0,0 +1,71 @@
+    def test_click_constructor(self, driver):
+        main_page = MainPage(driver)
+        main_page.open_page(Urls.BASE_URL)
+        main_page.click_to_element_with_js(MainFuncPageLocators.CONSTRUCTOR_BUTTON)
```
Нужно исправить: здесь и далее: локаторов в тестах быть не должно, если это не вариант параметризации, они живут в методах пейджи, тест должен быть максимально изолирован от логики выполнения

RealMadSquirrel replied:
Поправила

Commit:
Исправления по комментариям от ревью:
 - добавлено в  фикстуру с генерацией данных создание пользователя + удаление пользователя
 - изменения в соответствующих тестах, где используются фикстуры
 - убрала закоменченый код
 - убрала создание объекта одной страницы в методах другой
 - убрала локатор из main_page.py
 - убраны локаторы из тестов, созданы соответствующие методы для каждой пейджы

AlexKitIt commented on locators/main_page_locators.py:
```diff
@@ -0,0 +1,29 @@
+    PERSONAL_BUTTON = (By.XPATH, "//a[@href = '/account']") # ссылка на Личный Кабинет
+    PROFILE_TITLE = (By.XPATH, "//a[@href = '/account/profile']")
+    LENTA_ORDERS = (By.XPATH, "//h1[text() = 'Лента заказов']")
+    LENTA_BUTTON = (By.XPATH, "//p[text() = 'Лента Заказов']/parent::a")
```
Повторный коммент: этот элемент есть и тут и в OrdersLentaPageLocators, такая же ситуация с CONSTRUCTOR_BUTTON в LoginPageLocators и ForgotPasswordPageLocators и тд. 
Необходимо пройтись по всем локаторам и убедиться, что элементы существуют в единственном экземпляре в каком-то одном объекте пейджи. Они не должны копипаститься. Если есть на всех страницах(по типу хедера) - то либо новую common пейджу или в main

RealMadSquirrel replied:
Убрала - https://github.com/RealMadSquirrel/Diplom_3/pull/1/commits/dc57ead3a74724e066dbb53073f3854e1b1022b1

AlexKitIt commented on pages/account_page.py:
```diff
@@ -0,0 +1,27 @@
+
+
+    @allure.step('Кликаем по кнопке «Выход»')
+    def click_restore_password(self):
+        self.wait_and_find_element(LoginPageLocators.FORGOT_PASSWORD)
+        self.click_to_element_with_js(LoginPageLocators.FORGOT_PASSWORD)
+        return ForgotPasswordPage(self.driver)
```
повторный коммент: Нужно исправить: здесь и далее: в методах пейджи описываются только методы работы с элементами на конкретной пейдже, работы с элементами с других страниц быть не должно 
Это AccountPage а элементы с LoginPageLocators, зачем то импортирован MainPageLocators
Кроме того - не создается один объект пейджи в методах другой. Так делать не надо, объекты стоит создавать в тестах, чтобы четко отслеживать по шагам где мы находимся

RealMadSquirrel replied:
Спасибо, торопилась, не убрала этот метод...в AccountPage метод click_restore_password не используется...поправила.

Commit:
Задание №3 Исправление по комментариям попытка №2:
 - прошла по всем локаторам каждой пейджы, распределила, чтобы локаторы существовали в единственном экземпляре в соответствующем объекте пейджи.
 - убрала в AccountPage метод click_restore_password, он там не использовался.
 - убрала  создание один объекта пейджи в методах другой. В прошлый раз просмотрела этот момент, думала везде поправила.

Commit:
Задание №3 Исправление по комментариям попытка №2:
 - прошла по всем локаторам каждой пейджы, распределила, чтобы локаторы существовали в единственном экземпляре в соответствующем объекте пейджи.
 - убрала в AccountPage метод click_restore_password, он там не использовался.
 - убрала  создание один объекта пейджи в методах другой. В прошлый раз просмотрела этот момент, думала везде поправила.

 + вместо метода check_exists_by_xpath прописала метод wait_and_find_element из base_page.py

Commit:
Задание №3 Исправление по комментариям попытка №2:
 - прошла по всем локаторам каждой пейджы, распределила, чтобы локаторы существовали в единственном экземпляре в соответствующем объекте пейджи.
 - убрала в AccountPage метод click_restore_password, он там не использовался.
 - убрала  создание один объекта пейджи в методах другой. В прошлый раз просмотрела этот момент, думала везде поправила.

 + вместо метода check_exists_by_xpath прописала метод wait_and_find_element из base_page.py
 + убрала метод check_invisible_cross, который нигде не используется
